numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1627,12 +1627,30 @@ _array_fromobject_generic(
                 goto finish;
             }
         }
-        /* One more chance */
+        /* One more chance for faster exit if user specified the dtype. */
         oldtype = PyArray_DESCR(oparr);
         if (PyArray_EquivTypes(oldtype, type)) {
             if (copy != NPY_COPY_ALWAYS && STRIDING_OK(oparr, order)) {
                 Py_INCREF(op);
-                ret = oparr;
+                if (oldtype == type) {
```

Comment:
It seems like either
* this operator is overly strict (or the wrong check), or
* this change fundamentally conflicts with the current notions of `is`

Reply:
Maybe this change is too deep. If a user says `asarray(a, dtype='i')`, are there members of https://numpy.org/devdocs/reference/c-api/types-and-structures.html#c.PyArray_Descr that might differ under ordinary circumstances? Might `hash` be uninitialized?

Reply:
I think this is right, and we just need to rethink the tests that are failing!